compiler/src/iree/compiler/Codegen/Common/GPU/GPUDistributeForall.cpp

```diff
@@ -0,0 +1,217 @@
+  if (!forallOp.getMapping()) {
+    return success();
+  }
+
+  ArrayAttr mapping = forallOp.getMappingAttr();
+  bool hasThreadMapping =
+      llvm::all_of(mapping, llvm::IsaPred<gpu::GPUThreadMappingAttr>);
+  bool hasWarpMapping =
+      llvm::all_of(mapping, llvm::IsaPred<gpu::GPUWarpMappingAttr>);
+
+  // Skip forall ops that are not mapped to GPU ids.
+  if (!hasThreadMapping && !hasWarpMapping) {
+    return success();
+  }
+
+  if (forallOp->getNumResults() != 0) {
+    forallOp.emitOpError("Cannot distribute scf.forall op on tensors.");
+    return failure();
+  }
+
+  if (!isDescendingRelativeMappingIndices(mapping.getValue())) {
+    forallOp.emitOpError("Cannot distribute forall op with non-descending "
+                         "relative iterator mapping");
+    return failure();
+  }
+
+  if (!llvm::all_of(mapping, [](Attribute attr) {
+        return cast<DeviceMappingAttrInterface>(attr).isLinearMapping();
+      })) {
+    forallOp.emitOpError("unimplemented: resolution of scf.forall ops without "
+                         "linear id mappings.");
+    return failure();
+  }
+
+  if (!forallOp.isNormalized()) {
+    forallOp.emitOpError("scf.forall op must be normalized for distribution.");
+    return failure();
+  }
+
+  MLIRContext *context = rewriter.getContext();
+  Location loc = forallOp.getLoc();
+  AffineExpr d0, d1;
+  bindDims(context, d0, d1);
+
+  // Divide the thread ID by the subgroup size if this loop is mapped to
+  // subgroups.
+  assert(!(hasThreadMapping && hasWarpMapping));
+  Value flatId = linearThreadId;
+  if (hasWarpMapping) {
+    OpFoldResult subgroupSizeVal = rewriter.getIndexAttr(subgroupSize);
+    flatId = affine::makeComposedAffineApply(rewriter, loc, d0.floorDiv(d1),
+                                             {flatId, subgroupSizeVal});
+  }
+
+  SmallVector<Value> delinSizes;
+  OpFoldResult totalLoopTripCount = rewriter.getIndexAttr(1);
+  for (auto workerCount : forallOp.getMixedUpperBound()) {
+    delinSizes.push_back(
+        getValueOrCreateConstantIndexOp(rewriter, loc, workerCount));
+    totalLoopTripCount = affine::makeComposedFoldedAffineApply(
+        rewriter, loc, d0 * d1, {totalLoopTripCount, workerCount});
+  }
+
+  int64_t flatTotalNumWorkers =
+      hasWarpMapping ? flatWorkgroupSize / subgroupSize : flatWorkgroupSize;
+  std::optional<int64_t> staticProducerCount =
+      getConstantIntValue(totalLoopTripCount);
+  bool perfectlyDivides =
+      staticProducerCount &&
+      staticProducerCount.value() % flatTotalNumWorkers == 0;
+
+  // Step 3. Create the `scf.for` loop for the loop.
+  // If the workgroup count perfectly divides the loop's worker count, then we
+  // can use a lower bound of 0 and keep the loop bounds static. This helps
+  // simplify later loop folding patterns without an `affine.linearize_index` op
+  // to help with inferring int ranges.
+  Value lb = perfectlyDivides ? rewriter.create<arith::ConstantIndexOp>(loc, 0)
```

Comment:
I think the mixing of the perfectly divides and not case makes the code hard to follow. Could you do a normalization after the fact that just does this all the time?

Reply:
I commented about this in an earlier PR that I don't have immediately on hand, but the short answer is that I don't see a short term way to avoid the extra logic here and loop normalization doesn't solve the problem. The code here generates 2 possible paths based on `perfectlyDivides`
```
// if perfectly divides is true
scf.for %c0 to %wg_size step %loop_trip_count

// else
scf.for %thread_id to %wg_size step %loop_trip_count
```

if `wg_size == %loop_trip_count`, then this will be
```
// if perfectly divides is true
scf.for %c0 to %wg_size step %wg_size

// else
scf.for %thread_id to %wg_size step %wg_size
```
The first loop can be trivially folded away, while the second one requires proving that `0 <= thread_id < %wg_size`, and then writing a folder for `scf.for` that can query for such range information, which requires a non-trivial amount of work to do properly. Instead just adding a few extra lines of logic in the resolution pattern (here) seems simpler to me because that's the point when we can most easily directly compare the workgroup size against the loop trip count.

Also as a side benefit, loops with constant iteration bounds is easier to analyze for loop unrolling.

Reply:
ok... still has the smell of a premature optimization but I see why you need it.

@krzysz00  maybe something for range inference to handle.

Reply:
Might be good to have a chat about this at some point - especially since the inequalities here feel more like the value bounds stuff someone else did than the constant range stuff ... but perhaps if https://github.com/iree-org/iree/pull/17707 got pulled in some of this might rewrite away eventually?

Reply:
I think there are two parts here. 1 is #17707, which might be good enough on its own, but the other would be `affine.linearize_index` if the affine maps generated here are too unwieldy.